### PR TITLE
Created package sampling/ Added SampleByKey UDF

### PR DESCRIPTION
**Pull request by evionkim**

### CODE REVIEW

File: test/pig/datafu/test/pig/sampling/SamplingTests.java [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+    
+    test.runScript();
+    assertOutput(test, "sampled", 
+                 "(A4,B1,3)","(A4,B2,3)","(A4,B3,59)","(A4,B4,29)",
+                 "(A6,B2,3)","(A6,B2,55)","(A6,B3,1)",
+                 "(A9,B3,92)","(A9,B1,42)","(A9,B2,1)","(A9,B3,0)");
+  }
+
+  @Test
+  public void sampleByKeyTest2() throws Exception
```
Comment (matthayes) [inline test/pig/datafu/test/pig/sampling/SamplingTests.java]: i'd remove this since it's redundant with previous

**Comment by evionkim** [comment]
As pointed out, removed the redundant test.

**Comment by matthayes** [comment]
Instead of using Random to convert the object hash code into a "random" double, you can use code from below.  In this method I use SHA-1 to hash the integer and then convert a subset of the bytes to a double.  When I generate a histogram the result is uniformly distributed.  Shorter sequences like 1:1000 are also pretty well distributed.

```
MessageDigest hasher = MessageDigest.getInstance("sha-1");

ByteBuffer b = ByteBuffer.allocate(4+4);
ByteBuffer b2 = ByteBuffer.allocate(20);

FileOutputStream fos = new FileOutputStream("data.txt"); 
OutputStreamWriter out = new OutputStreamWriter(fos, "UTF-8");

int seed = 429524;

for (int i=0; i<1000000; i++)
{
  b.putInt(seed);
  b.putInt(i);
  byte[] digest = hasher.digest(b.array());
  b.clear();

  b2.put(digest);
  b2.rewind();
  double result = (((double)b2.getInt())/Integer.MAX_VALUE  + 1)/2;
  b2.clear();

  out.write(Double.toString(result) + "\n");
}    

out.close();
fos.close();
```

**Comment by matthayes** [comment]
Here are some histograms from the code I just shared.

1 through 100:

data100

1 through 1000:

data1000

1 through 10,000:

data10000

1 through 1,000,000:

data1000000

**Comment by evionkim** [comment]
Random double generate algorithm now uses SHA-1 based method. Changed tests accordingly. Also, added reservoir sample UDF.

**Pull request merged**